php/EE/Migration/Containers.php

```diff
@@ -0,0 +1,150 @@
+
+					default_launch( 'docker rm -f ee-nginx-proxy', false, true );
+
+					if ( ! default_launch( $ee_proxy_command, false, true ) ) {
+						throw new \Exception( ' Unable to restore ee-nginx-proxy container' );
+					}
+				}
+			);
+		}
+
+		// Upgrade cron container
+		$existing_cron_image = EE::launch( 'docker inspect --format=\'{{.Config.Image}}\' ee-cron-scheduler', false, true );
+		if ( $existing_cron_image->return_code === 0 ) {
+			self::$rsp->execute_step(
+				function () {
+					$cron_image = 'easyengine/cron:v' . EE_VERSION;
+					$cron_scheduler_run_command = 'docker run --name ee-cron-scheduler --restart=always -d -v ' . EE_CONF_ROOT . '/cron:/etc/ofelia:ro -v /var/run/docker.sock:/var/run/docker.sock:ro ' . $cron_image ;
+
+					default_launch( 'docker rm -f ee-cron-scheduler', false, true );
+
+					if ( ! default_launch( $cron_scheduler_run_command, false, true ) ) {
+						throw new \Exception( ' Unable to upgrade ee-cron-scheduler container' );
+					}
+				},
+				function () use ( $existing_cron_image ) {
+					$cron_image = trim( $existing_cron_image->stdout );
+					$cron_scheduler_run_command = 'docker run --name ee-cron-scheduler --restart=always -d -v ' . EE_CONF_ROOT . '/cron:/etc/ofelia:ro -v /var/run/docker.sock:/var/run/docker.sock:ro ' . $cron_image ;
+
+					default_launch( 'docker rm -f ee-cron-scheduler', false, true );
+
+					if ( ! default_launch( $cron_scheduler_run_command, false, true ) ) {
+						throw new \Exception( ' Unable to restore ee-cron-scheduler container' );
+					}
+				}
+			);
+		}
+	}
+
+	/**
+	 * Migrates all containers of existing sites
+	 */
+	private static function migrate_site_containers() {
+		$sites       = \EE_DB::select( [ 'sitename', 'site_path', 'site_type', 'cache_type', 'is_ssl', 'db_host' ] );
+		$site_docker = new \Site_Docker();
+
+		foreach ( $sites as $site ) {
+
+			$data[] = $site['site_type'];
+			$data[] = $site['cache_type'];
+			$data[] = $site['is_ssl'];
+			$data[] = $site['db_host'];
+
+			$docker_compose_contents = $site_docker->generate_docker_compose_yml( $data );
+			$docker_compose_path     = $site['site_path'] . '/docker-compose.yml';
+			$docker_compose_backup_path     = $site['site_path'] . '/docker-compose.yml.bak';
+
+			self::$rsp->execute_step(
+				function () use ( $site, $docker_compose_path, $docker_compose_backup_path ) {
+					if ( ! default_launch( "cp $docker_compose_path $docker_compose_backup_path" ) ) {
+						throw new \Exception( "Unable to find docker-compose.yml in ${site['site_path']} or couldn't create it's backup file. Ensure that EasyEngine has permission to create file there" );
+					}
+				}, function () {}
```

Comment:
You're not using $context here and you're using it in o/p messages w/o checking null/empty. I think we should make it required so that it can help in debugging the issue when we face one.